src/op.rs

```diff
@@ -1198,34 +1266,210 @@ impl<'context, 'input, Endian, Context> Evaluation<'context, 'input, Endian, Con
+            terminated: terminated,
+            piece_end: piece_end,
+            current_location: current_location,
+        })
     }
 
-    /// Evaluate a DWARF expression.  If successful, the result will
-    /// hold a vector of pieces, each describing a part of the final
-    /// object.
-    pub fn evaluate(&mut self) -> Result<Vec<Piece<'input>>, Context::ContextError>
-        where Endian: Endianity
+    /// Get the evaluation result.  This `Evaluation` must have previously been
+    /// driven to completion by calling the evaluate and resume_with methods
+    /// until `EvaluationResult::Complete` was returned.
+    pub fn result(self) -> Vec<Piece<'input>> {
+        match self.state {
+            EvaluationState::Complete => self.result,
+            _ => panic!(),
+        }
+    }
+
+    /// Evaluate a DWARF expression.  This method should only ever be called
+    /// once.  If the returned `EvaluationResult` is not
+    /// `EvaluationResult::Complete`, the caller should provide the required
+    /// value and resume the evaluation by calling one of the resume_with
+    /// methods on Evaluation.
+    pub fn evaluate(&mut self) -> Result<EvaluationResult<'input, Endian>, Error>
+        where Endian: Endianity,
     {
-        if let Some(value) = self.initial_value {
-            self.push(value);
+        match self.state {
+            EvaluationState::Start(initial_value) => {
+                if let Some(value) = initial_value {
+                    self.push(value);
+                }
+                self.state = EvaluationState::Ready;
+            },
+            EvaluationState::Ready => {},
+            EvaluationState::Error(err) => return Err(err),
+            EvaluationState::Complete => return Ok(EvaluationResult::Complete),
+            EvaluationState::Waiting(_) => panic!(),
+        };
+
+        match self.evaluate_internal() {
+            Ok(r) => Ok(r),
+            Err(e) => {
+                self.state = EvaluationState::Error(e);
+                Err(e)
+            },
         }
+    }
 
-        // The number of instructions we've processed.
-        let mut iteration = 0;
+    /// Resume the `Evaluation` with the provided `value`.  A call to this
+    /// function must follow `Evaluation::evaluate` or one of the resume_with
+    /// functions returning `EvaluationResult::RequiresMemory`.  This will apply
+    /// the provided memory value to the evaluation and continue evaluating
+    /// opcodes until the evaluation is completed, reaches an error, or needs
+    /// more information again.
+    pub fn resume_with_memory(&mut self, value: u64) -> Result<EvaluationResult<'input, Endian>, Error>
+        where Endian: Endianity
+    {
+        match self.state {
+            EvaluationState::Error(err) => return Err(err),
+            EvaluationState::Waiting(OperationEvaluationResult::AwaitingMemory { .. }) => {
+                self.push(value);
+            },
+            _ => panic!(),
```

Comment:
Ditto with regards to a diagnostic message for the panic and a section in the doc comment.